Setup shared cluster on AWS and deploy 'researchdelight' hub

sgibson91 opened the pull request:
ref 2i2c-org/infrastructure#1949 

Since this is a new shared cluster, I will deploy a staging and dask-staging hub alongside the research delight hub requested, mimicking the GCP shared cluster setup.

Hubs added:

- https://staging.aws.2i2c.cloud
- https://dask-staging.aws.2i2c.cloud
- https://researchdelight.2i2c.cloud

sgibson91 commented:
tf plan output:

```
Terraform used the selected providers to generate the following execution plan. Resource actions are indicated with the following symbols:
  + create
 <= read (data resources)

Terraform will perform the following actions:

  # data.aws_iam_policy_document.bucket_access["prod.scratch"] will be read during apply
  # (config refers to values not yet known)
 <= data "aws_iam_policy_document" "bucket_access" {
      + id   = (known after apply)
      + json = (known after apply)

      + statement {
          + actions   = [
              + "s3:*",
            ]
          + effect    = "Allow"
          + resources = [
              + (known after apply),
              + (known after apply),
            ]

          + principals {
              + identifiers = [
                  + (known after apply),
                ]
              + type        = "AWS"
            }
        }
    }

  # data.aws_iam_policy_document.bucket_access["staging.scratch-staging"] will be read during apply
  # (config refers to values not yet known)
 <= data "aws_iam_policy_document" "bucket_access" {
      + id   = (known after apply)
      + json = (known after apply)

      + statement {
          + actions   = [
              + "s3:*",
            ]
          + effect    = "Allow"
          + resources = [
              + (known after apply),
              + (known after apply),
            ]

          + principals {
              + identifiers = [
                  + (known after apply),
                ]
              + type        = "AWS"
            }
        }
    }

  # aws_efs_file_system.homedirs will be created
  + resource "aws_efs_file_system" "homedirs" {
      + arn                     = (known after apply)
      + availability_zone_id    = (known after apply)
      + availability_zone_name  = (known after apply)
      + creation_token          = (known after apply)
      + dns_name                = (known after apply)
      + encrypted               = (known after apply)
      + id                      = (known after apply)
      + kms_key_id              = (known after apply)
      + number_of_mount_targets = (known after apply)
      + owner_id                = (known after apply)
      + performance_mode        = (known after apply)
      + size_in_bytes           = (known after apply)
      + tags                    = {
          + "Name" = "hub-homedirs"
        }
      + tags_all                = {
          + "Name" = "hub-homedirs"
        }
      + throughput_mode         = "bursting"
    }

  # aws_efs_mount_target.homedirs will be created
  + resource "aws_efs_mount_target" "homedirs" {
      + availability_zone_id   = (known after apply)
      + availability_zone_name = (known after apply)
      + dns_name               = (known after apply)
      + file_system_arn        = (known after apply)
      + file_system_id         = (known after apply)
      + id                     = (known after apply)
      + ip_address             = (known after apply)
      + mount_target_dns_name  = (known after apply)
      + network_interface_id   = (known after apply)
      + owner_id               = (known after apply)
      + security_groups        = [
          + "sg-05bde4ef8b3ba3976",
        ]
      + subnet_id              = "subnet-0ecfb9634dcce8c41"
    }

  # aws_iam_access_key.continuous_deployer will be created
  + resource "aws_iam_access_key" "continuous_deployer" {
      + create_date                    = (known after apply)
      + encrypted_secret               = (known after apply)
      + encrypted_ses_smtp_password_v4 = (known after apply)
      + id                             = (known after apply)
      + key_fingerprint                = (known after apply)
      + secret                         = (sensitive value)
      + ses_smtp_password_v4           = (sensitive value)
      + status                         = "Active"
      + user                           = "hub-continuous-deployer"
    }

  # aws_iam_role.irsa_role["prod"] will be created
  + resource "aws_iam_role" "irsa_role" {
      + arn                   = (known after apply)
      + assume_role_policy    = jsonencode(
            {
              + Statement = [
                  + {
                      + Action    = "sts:AssumeRoleWithWebIdentity"
                      + Condition = {
                          + StringEquals = {
                              + "oidc.eks.us-west-2.amazonaws.com/id/87AF0BEB166A8A86CFB4EF059AC41C1B:sub" = "system:serviceaccount:prod:user-sa"
                            }
                        }
                      + Effect    = "Allow"
                      + Principal = {
                          + Federated = "arn:aws:iam::790657130469:oidc-provider/oidc.eks.us-west-2.amazonaws.com/id/87AF0BEB166A8A86CFB4EF059AC41C1B"
                        }
                      + Sid       = ""
                    },
                ]
              + Version   = "2012-10-17"
            }
        )
      + create_date           = (known after apply)
      + force_detach_policies = false
      + id                    = (known after apply)
      + managed_policy_arns   = (known after apply)
      + max_session_duration  = 3600
      + name                  = "shared-hubs-cluster-prod"
      + name_prefix           = (known after apply)
      + path                  = "/"
      + tags_all              = (known after apply)
      + unique_id             = (known after apply)

      + inline_policy {
          + name   = (known after apply)
          + policy = (known after apply)
        }
    }

  # aws_iam_role.irsa_role["staging"] will be created
  + resource "aws_iam_role" "irsa_role" {
      + arn                   = (known after apply)
      + assume_role_policy    = jsonencode(
            {
              + Statement = [
                  + {
                      + Action    = "sts:AssumeRoleWithWebIdentity"
                      + Condition = {
                          + StringEquals = {
                              + "oidc.eks.us-west-2.amazonaws.com/id/87AF0BEB166A8A86CFB4EF059AC41C1B:sub" = "system:serviceaccount:staging:user-sa"
                            }
                        }
                      + Effect    = "Allow"
                      + Principal = {
                          + Federated = "arn:aws:iam::790657130469:oidc-provider/oidc.eks.us-west-2.amazonaws.com/id/87AF0BEB166A8A86CFB4EF059AC41C1B"
                        }
                      + Sid       = ""
                    },
                ]
              + Version   = "2012-10-17"
            }
        )
      + create_date           = (known after apply)
      + force_detach_policies = false
      + id                    = (known after apply)
      + managed_policy_arns   = (known after apply)
      + max_session_duration  = 3600
      + name                  = "shared-hubs-cluster-staging"
      + name_prefix           = (known after apply)
      + path                  = "/"
      + tags_all              = (known after apply)
      + unique_id             = (known after apply)

      + inline_policy {
          + name   = (known after apply)
          + policy = (known after apply)
        }
    }

  # aws_iam_user.continuous_deployer will be created
  + resource "aws_iam_user" "continuous_deployer" {
      + arn           = (known after apply)
      + force_destroy = false
      + id            = (known after apply)
      + name          = "hub-continuous-deployer"
      + path          = "/"
      + tags_all      = (known after apply)
      + unique_id     = (known after apply)
    }

  # aws_iam_user_policy.continuous_deployer will be created
  + resource "aws_iam_user_policy" "continuous_deployer" {
      + id     = (known after apply)
      + name   = "eks-readonly"
      + policy = jsonencode(
            {
              + Statement = [
                  + {
                      + Action   = "eks:DescribeCluster"
                      + Effect   = "Allow"
                      + Resource = "*"
                    },
                ]
              + Version   = "2012-10-17"
            }
        )
      + user   = "hub-continuous-deployer"
    }

  # aws_s3_bucket.user_buckets["scratch"] will be created
  + resource "aws_s3_bucket" "user_buckets" {
      + acceleration_status         = (known after apply)
      + acl                         = (known after apply)
      + arn                         = (known after apply)
      + bucket                      = "shared-hubs-cluster-scratch"
      + bucket_domain_name          = (known after apply)
      + bucket_regional_domain_name = (known after apply)
      + force_destroy               = false
      + hosted_zone_id              = (known after apply)
      + id                          = (known after apply)
      + object_lock_enabled         = (known after apply)
      + policy                      = (known after apply)
      + region                      = (known after apply)
      + request_payer               = (known after apply)
      + tags_all                    = (known after apply)
      + website_domain              = (known after apply)
      + website_endpoint            = (known after apply)

      + cors_rule {
          + allowed_headers = (known after apply)
          + allowed_methods = (known after apply)
          + allowed_origins = (known after apply)
          + expose_headers  = (known after apply)
          + max_age_seconds = (known after apply)
        }

      + grant {
          + id          = (known after apply)
          + permissions = (known after apply)
          + type        = (known after apply)
          + uri         = (known after apply)
        }

      + lifecycle_rule {
          + abort_incomplete_multipart_upload_days = (known after apply)
          + enabled                                = (known after apply)
          + id                                     = (known after apply)
          + prefix                                 = (known after apply)
          + tags                                   = (known after apply)

          + expiration {
              + date                         = (known after apply)
              + days                         = (known after apply)
              + expired_object_delete_marker = (known after apply)
            }

          + noncurrent_version_expiration {
              + days = (known after apply)
            }

          + noncurrent_version_transition {
              + days          = (known after apply)
              + storage_class = (known after apply)
            }

          + transition {
              + date          = (known after apply)
              + days          = (known after apply)
              + storage_class = (known after apply)
            }
        }

      + logging {
          + target_bucket = (known after apply)
          + target_prefix = (known after apply)
        }

      + object_lock_configuration {
          + object_lock_enabled = (known after apply)

          + rule {
              + default_retention {
                  + days  = (known after apply)
                  + mode  = (known after apply)
                  + years = (known after apply)
                }
            }
        }

      + replication_configuration {
          + role = (known after apply)

          + rules {
              + delete_marker_replication_status = (known after apply)
              + id                               = (known after apply)
              + prefix                           = (known after apply)
              + priority                         = (known after apply)
              + status                           = (known after apply)

              + destination {
                  + account_id         = (known after apply)
                  + bucket             = (known after apply)
                  + replica_kms_key_id = (known after apply)
                  + storage_class      = (known after apply)

                  + access_control_translation {
                      + owner = (known after apply)
                    }

                  + metrics {
                      + minutes = (known after apply)
                      + status  = (known after apply)
                    }

                  + replication_time {
                      + minutes = (known after apply)
                      + status  = (known after apply)
                    }
                }

              + filter {
                  + prefix = (known after apply)
                  + tags   = (known after apply)
                }

              + source_selection_criteria {
                  + sse_kms_encrypted_objects {
                      + enabled = (known after apply)
                    }
                }
            }
        }

      + server_side_encryption_configuration {
          + rule {
              + bucket_key_enabled = (known after apply)

              + apply_server_side_encryption_by_default {
                  + kms_master_key_id = (known after apply)
                  + sse_algorithm     = (known after apply)
                }
            }
        }

      + versioning {
          + enabled    = (known after apply)
          + mfa_delete = (known after apply)
        }

      + website {
          + error_document           = (known after apply)
          + index_document           = (known after apply)
          + redirect_all_requests_to = (known after apply)
          + routing_rules            = (known after apply)
        }
    }

  # aws_s3_bucket.user_buckets["scratch-staging"] will be created
  + resource "aws_s3_bucket" "user_buckets" {
      + acceleration_status         = (known after apply)
      + acl                         = (known after apply)
      + arn                         = (known after apply)
      + bucket                      = "shared-hubs-cluster-scratch-staging"
      + bucket_domain_name          = (known after apply)
      + bucket_regional_domain_name = (known after apply)
      + force_destroy               = false
      + hosted_zone_id              = (known after apply)
      + id                          = (known after apply)
      + object_lock_enabled         = (known after apply)
      + policy                      = (known after apply)
      + region                      = (known after apply)
      + request_payer               = (known after apply)
      + tags_all                    = (known after apply)
      + website_domain              = (known after apply)
      + website_endpoint            = (known after apply)

      + cors_rule {
          + allowed_headers = (known after apply)
          + allowed_methods = (known after apply)
          + allowed_origins = (known after apply)
          + expose_headers  = (known after apply)
          + max_age_seconds = (known after apply)
        }

      + grant {
          + id          = (known after apply)
          + permissions = (known after apply)
          + type        = (known after apply)
          + uri         = (known after apply)
        }

      + lifecycle_rule {
          + abort_incomplete_multipart_upload_days = (known after apply)
          + enabled                                = (known after apply)
          + id                                     = (known after apply)
          + prefix                                 = (known after apply)
          + tags                                   = (known after apply)

          + expiration {
              + date                         = (known after apply)
              + days                         = (known after apply)
              + expired_object_delete_marker = (known after apply)
            }

          + noncurrent_version_expiration {
              + days = (known after apply)
            }

          + noncurrent_version_transition {
              + days          = (known after apply)
              + storage_class = (known after apply)
            }

          + transition {
              + date          = (known after apply)
              + days          = (known after apply)
              + storage_class = (known after apply)
            }
        }

      + logging {
          + target_bucket = (known after apply)
          + target_prefix = (known after apply)
        }

      + object_lock_configuration {
          + object_lock_enabled = (known after apply)

          + rule {
              + default_retention {
                  + days  = (known after apply)
                  + mode  = (known after apply)
                  + years = (known after apply)
                }
            }
        }

      + replication_configuration {
          + role = (known after apply)

          + rules {
              + delete_marker_replication_status = (known after apply)
              + id                               = (known after apply)
              + prefix                           = (known after apply)
              + priority                         = (known after apply)
              + status                           = (known after apply)

              + destination {
                  + account_id         = (known after apply)
                  + bucket             = (known after apply)
                  + replica_kms_key_id = (known after apply)
                  + storage_class      = (known after apply)

                  + access_control_translation {
                      + owner = (known after apply)
                    }

                  + metrics {
                      + minutes = (known after apply)
                      + status  = (known after apply)
                    }

                  + replication_time {
                      + minutes = (known after apply)
                      + status  = (known after apply)
                    }
                }

              + filter {
                  + prefix = (known after apply)
                  + tags   = (known after apply)
                }

              + source_selection_criteria {
                  + sse_kms_encrypted_objects {
                      + enabled = (known after apply)
                    }
                }
            }
        }

      + server_side_encryption_configuration {
          + rule {
              + bucket_key_enabled = (known after apply)

              + apply_server_side_encryption_by_default {
                  + kms_master_key_id = (known after apply)
                  + sse_algorithm     = (known after apply)
                }
            }
        }

      + versioning {
          + enabled    = (known after apply)
          + mfa_delete = (known after apply)
        }

      + website {
          + error_document           = (known after apply)
          + index_document           = (known after apply)
          + redirect_all_requests_to = (known after apply)
          + routing_rules            = (known after apply)
        }
    }

  # aws_s3_bucket_lifecycle_configuration.user_bucket_expiry["scratch"] will be created
  + resource "aws_s3_bucket_lifecycle_configuration" "user_bucket_expiry" {
      + bucket = "shared-hubs-cluster-scratch"
      + id     = (known after apply)

      + rule {
          + id     = "delete-after-expiry"
          + status = "Enabled"

          + expiration {
              + days                         = 7
              + expired_object_delete_marker = (known after apply)
            }
        }
    }

  # aws_s3_bucket_lifecycle_configuration.user_bucket_expiry["scratch-staging"] will be created
  + resource "aws_s3_bucket_lifecycle_configuration" "user_bucket_expiry" {
      + bucket = "shared-hubs-cluster-scratch-staging"
      + id     = (known after apply)

      + rule {
          + id     = "delete-after-expiry"
          + status = "Enabled"

          + expiration {
              + days                         = 7
              + expired_object_delete_marker = (known after apply)
            }
        }
    }

  # aws_s3_bucket_policy.user_bucket_access["prod.scratch"] will be created
  + resource "aws_s3_bucket_policy" "user_bucket_access" {
      + bucket = (known after apply)
      + id     = (known after apply)
      + policy = (known after apply)
    }

  # aws_s3_bucket_policy.user_bucket_access["staging.scratch-staging"] will be created
  + resource "aws_s3_bucket_policy" "user_bucket_access" {
      + bucket = (known after apply)
      + id     = (known after apply)
      + policy = (known after apply)
    }

Plan: 13 to add, 0 to change, 0 to destroy.

Changes to Outputs:
  + continuous_deployer_creds = (sensitive value)
  + db_helm_config            = (sensitive value)
  + eksctl_iam_command        = (known after apply)
  + kubernetes_sa_annotations = {
      + prod    = (known after apply)
      + staging = (known after apply)
    }
  + nfs_server_dns            = (known after apply)
```

sgibson91 commented:
@2i2c-org/tech-team I am requesting early review of the eksctl/terraform files before I fully deploy and begin on the hubs

yuvipanda commented on eksctl/shared-hubs-cluster.jsonnet:
```diff
@@ -0,0 +1,112 @@
+        // Warning: version 1.23 introduces some breaking changes
+        // Checkout the docs before upgrading
+        // ref: https://docs.aws.amazon.com/eks/latest/userguide/ebs-csi-migration-faq.html
+        version: '1.22'
```
Explicitly stating that this is ok for now to get this out the door, and we should work on upgrading in the new year.

sgibson91 replied:
Yes! Scale to zero node groups!

yuvipanda reviewed (commented):
LGTM @sgibson91. The only suggestion I have is that we name this to match the other shared projects we have, and call it `2i2c-aws` or `2i2c-aws-us` maybe? I think `shared-cluster` is a bit too broad.

sgibson91 commented:
> The only suggestion I have is that we name this to match the other shared projects we have, and call it `2i2c-aws` or `2i2c-aws-us` maybe? I think `shared-cluster` is a bit too broad.

Yeah, I was trying to look to the two-eye-two-see GCP project for inspiration but that's still called `pilot-hubs-cluster` 😬

sgibson91 commented:
Latest commit renames the cluster to `2i2c-aws-us`

sgibson91 commented:
I just deployed a staging hub to this cluster more or less copying the staging hub from the 2i2c/GCP shared cluster folder.

* https://staging.aws.2i2c.cloud

It deployed fine, but I am not an admin when I login? Even though I have provided the correct config to make me an admin

- https://github.com/2i2c-org/infrastructure/pull/1967/files#diff-308ebe1f478ae7d2f6bda31ca7db832af923e652095d35b12641619804226464R17-R18
- https://github.com/2i2c-org/infrastructure/pull/1967/files#diff-72f80ce5822fbdb4dc2c9b1983c569890c3b828898052ef868c5ef15f2f97eb5R14-R17

<img width="955" alt="Screenshot 2022-12-02 at 17 49 48" src="https://user-images.githubusercontent.com/44771837/205355170-8dd62170-06c8-44c9-ab72-7977327198b8.png">

yuvipanda commented:
@sgibson91 what does the hub logs say?

sgibson91 commented:
> what does the hub logs say?

Nothing interesting at all. The config to make us admins runs without error.

sgibson91 commented:
There's this?

```
[W 2022-12-02 17:49:28.048 JupyterHub auth:298] Service Server at /user/sgibson@2i2c.org/ requested scopes access:servers!server=sgibson@2i2c.org/,access:servers!user=sgibson@2i2c.org,read:users:name!user,read:users:groups!user for user sgibson@2i2c.org, granting only access:servers!server=sgibson@2i2c.org/,read:users:name!user,read:users:groups!user.
```

sgibson91 commented:
Initial start up logs:

```
Loading /usr/local/etc/jupyterhub/secret/values.yaml
No config at /usr/local/etc/jupyterhub/existing-secret/values.yaml
Loading extra config: 01-custom-theme
Loading extra config: 02-custom-admin
Loading extra config: 03-cloud-storage-bucket
Loading extra config: 04-2i2c-add-staff-user-ids-to-admin-users
Loading extra config: 05-gh-teams
Loading extra config: 05-per-user-disk
[I 2022-12-02 17:38:51.717 JupyterHub app:2775] Running JupyterHub version 3.0.0
[I 2022-12-02 17:38:51.717 JupyterHub app:2805] Using Authenticator: oauthenticator.auth0.Auth0OAuthenticator-15.1.0
[I 2022-12-02 17:38:51.717 JupyterHub app:2805] Using Spawner: builtins.CustomSpawner
[I 2022-12-02 17:38:51.717 JupyterHub app:2805] Using Proxy: jupyterhub.proxy.ConfigurableHTTPProxy-3.0.0
[I 2022-12-02 17:38:51.788 JupyterHub app:1934] Not using allowed_users. Any authenticated user will be allowed.
[I 2022-12-02 17:38:51.815 JupyterHub provider:653] Updating oauth client service-configurator
[I 2022-12-02 17:38:51.871 JupyterHub app:2844] Initialized 0 spawners in 0.003 seconds
[I 2022-12-02 17:38:51.872 JupyterHub app:3057] Not starting proxy
[I 2022-12-02 17:38:51.876 JupyterHub app:3093] Hub API listening on http://:8081/hub/
[I 2022-12-02 17:38:51.876 JupyterHub app:3095] Private Hub API connect url http://hub:8081/hub/
[I 2022-12-02 17:38:51.877 JupyterHub app:3104] Starting managed service jupyterhub-idle-culler
[I 2022-12-02 17:38:51.877 JupyterHub service:385] Starting service 'jupyterhub-idle-culler': ['python3', '-m', 'jupyterhub_idle_culler', '--url=http://localhost:8081/hub/api', '--timeout=3600', '--cull-every=600', '--concurrency=10']
[I 2022-12-02 17:38:51.879 JupyterHub service:133] Spawning python3 -m jupyterhub_idle_culler --url=http://localhost:8081/hub/api --timeout=3600 --cull-every=600 --concurrency=10
[I 2022-12-02 17:38:51.883 JupyterHub app:3104] Starting managed service configurator at http://configurator:10101
[I 2022-12-02 17:38:51.883 JupyterHub service:385] Starting service 'configurator': ['python3', '-m', 'jupyterhub_configurator.app', '--Configurator.config_file=/usr/local/etc/jupyterhub-configurator/jupyterhub_configurator_config.py']
[I 2022-12-02 17:38:51.885 JupyterHub service:133] Spawning python3 -m jupyterhub_configurator.app --Configurator.config_file=/usr/local/etc/jupyterhub-configurator/jupyterhub_configurator_config.py
[I 2022-12-02 17:38:52.040 JupyterHub log:186] 200 GET /hub/api/ (jupyterhub-idle-culler@127.0.0.1) 11.61ms
[I 2022-12-02 17:38:52.054 JupyterHub log:186] 200 GET /hub/api/users?state=[secret] (jupyterhub-idle-culler@127.0.0.1) 12.19ms
[I 2022-12-02 17:38:52.899 JupyterHub app:3113] Adding external service hub-health
[I 2022-12-02 17:38:52.900 JupyterHub proxy:480] Adding route for Hub: / => http://hub:8081
[W 2022-12-02 17:38:52.901 JupyterHub proxy:448] Adding missing route for configurator (Server(url=http://configurator:10101/services/configurator/, bind_url=http://configurator:10101/services/configurator/))
[I 2022-12-02 17:38:52.902 JupyterHub proxy:314] Adding service configurator to proxy /services/configurator/ => http://configurator:10101
[I 2022-12-02 17:38:52.906 JupyterHub app:3162] JupyterHub is now running, internal Hub API at http://hub:8081/hub/
```

sgibson91 commented:
From when I logged in

```
[I 2022-12-02 17:41:46.867 JupyterHub oauth2:102] OAuth redirect: 'https://staging.aws.2i2c.cloud/hub/oauth_callback'
[I 2022-12-02 17:41:46.868 JupyterHub log:186] 302 GET /hub/oauth_login?next=%2Fhub%2F -> https://2i2c.us.auth0.com/authorize?response_type=code&redirect_uri=https%3A%2F%2Fstaging.aws.2i2c.cloud%2Fhub%2Foauth_callback&client_id=nTFsGt6LL2SPwLuQEsu5krQpUp8Xr18a&state=[secret]&scope=openid+name+profile+email (@192.168.6.151) 1.67ms
[I 2022-12-02 17:41:55.466 JupyterHub roles:238] Adding role user for User: sgibson@2i2c.org
[I 2022-12-02 17:41:55.482 JupyterHub base:810] User logged in: sgibson@2i2c.org
[I 2022-12-02 17:41:55.482 JupyterHub log:186] 302 GET /hub/oauth_callback?code=[secret]&state=[secret] -> /hub/ (@192.168.6.151) 397.10ms
[I 2022-12-02 17:41:55.774 JupyterHub log:186] 302 GET /hub/ -> /hub/spawn (sgibson@2i2c.org@192.168.6.151) 15.73ms
[I 2022-12-02 17:41:55.815 JupyterHub reflector:274] watching for pods with label selector='component=singleuser-server' in namespace staging
[I 2022-12-02 17:41:55.818 JupyterHub reflector:274] watching for events with field selector='involvedObject.kind=Pod' in namespace staging
[I 2022-12-02 17:41:55.987 JupyterHub provider:651] Creating oauth client jupyterhub-user-sgibson%402i2c.org
[I 2022-12-02 17:41:56.006 JupyterHub log:186] 302 GET /hub/spawn -> /hub/spawn-pending/sgibson@2i2c.org (sgibson@2i2c.org@192.168.6.151) 32.70ms
[I 2022-12-02 17:41:56.015 JupyterHub spawner:2489] Attempting to create pod jupyter-sgibson-402i2c-2eorg, with timeout 3
[I 2022-12-02 17:41:56.201 JupyterHub pages:394] sgibson@2i2c.org is pending spawn
[I 2022-12-02 17:41:56.204 JupyterHub log:186] 200 GET /hub/spawn-pending/sgibson@2i2c.org (sgibson@2i2c.org@192.168.6.151) 5.86ms
[I 2022-12-02 17:42:07.963 JupyterHub log:186] 200 GET /hub/home (sgibson@2i2c.org@192.168.6.151) 11.55ms
[I 2022-12-02 17:42:14.758 JupyterHub log:186] 302 GET / -> /hub/ (@192.168.6.151) 0.72ms
[I 2022-12-02 17:42:14.909 JupyterHub log:186] 302 GET /hub/ -> /hub/login?next=%2Fhub%2F (@192.168.6.151) 0.72ms
[I 2022-12-02 17:42:15.062 JupyterHub log:186] 200 GET /hub/login?next=%2Fhub%2F (@192.168.6.151) 1.96ms
[W 2022-12-02 17:42:15.310 JupyterHub web:1796] 400 DELETE /hub/api/users/sgibson%402i2c.org/server (192.168.6.151): sgibson@2i2c.org is pending spawn, please wait
[W 2022-12-02 17:42:15.311 JupyterHub log:186] 400 DELETE /hub/api/users/sgibson%402i2c.org/server (sgibson@2i2c.org@192.168.6.151) 2.93ms
[I 2022-12-02 17:42:24.216 JupyterHub log:186] 200 GET /hub/metrics (@192.168.8.103) 6.03ms
[I 2022-12-02 17:43:24.216 JupyterHub log:186] 200 GET /hub/metrics (@192.168.8.103) 6.03ms
[I 2022-12-02 17:44:24.217 JupyterHub log:186] 200 GET /hub/metrics (@192.168.8.103) 6.96ms
[I 2022-12-02 17:44:58.610 JupyterHub log:186] 301 GET /user/sgibson@2i2c.org -> /user/sgibson@2i2c.org/ (@192.168.6.151) 0.82ms
[I 2022-12-02 17:44:58.981 JupyterHub log:186] 302 GET /user/sgibson@2i2c.org/ -> /hub/user/sgibson@2i2c.org/ (@192.168.6.151) 0.79ms
[I 2022-12-02 17:44:59.167 JupyterHub log:186] 303 GET /hub/user/sgibson@2i2c.org/ (sgibson@2i2c.org@192.168.6.151) 7.58ms
[I 2022-12-02 17:44:59.348 JupyterHub pages:394] sgibson@2i2c.org is pending spawn
[I 2022-12-02 17:44:59.349 JupyterHub log:186] 200 GET /hub/spawn-pending/sgibson@2i2c.org?next=%2Fhub%2Fuser%2Fsgibson%402i2c.org%2F (sgibson@2i2c.org@192.168.6.151) 3.58ms
[I 2022-12-02 17:45:03.044 JupyterHub log:186] 200 GET /hub/api (@192.168.6.204) 0.62ms
[I 2022-12-02 17:45:03.065 JupyterHub log:186] 200 POST /hub/api/users/sgibson@2i2c.org/activity (sgibson@2i2c.org@192.168.6.204) 14.62ms
[W 2022-12-02 17:45:03.312 JupyterHub _version:68] jupyterhub version 3.0.0 != jupyterhub-singleuser version 2.3.1. This could cause failure to authenticate and result in redirect loops!
[I 2022-12-02 17:45:03.312 JupyterHub base:963] User sgibson@2i2c.org took 187.337 seconds to start
[I 2022-12-02 17:45:03.312 JupyterHub proxy:333] Adding user sgibson@2i2c.org to proxy /user/sgibson@2i2c.org/ => http://192.168.6.204:8888
[I 2022-12-02 17:45:03.314 JupyterHub users:749] Server sgibson@2i2c.org is ready
```

yuvipanda commented:
My suggestion is to put some print statements here https://github.com/2i2c-org/infrastructure/blob/master/helm-charts/basehub/values.yaml#L492 and see what shows up in the hub logs.

sgibson91 commented:
Script after adding print statements:

```
from z2jh import get_config

print("We are adding staff IDs to the hub admin list!")

add_staff_user_ids_to_admin_users = get_config("custom.2i2c.add_staff_user_ids_to_admin_users", False)
print(f"Are we adding?: {add_staff_user_ids_to_admin_users}")

if add_staff_user_ids_to_admin_users:
    print("We made it to the if statement!")

    user_id_type = get_config("custom.2i2c.add_staff_user_ids_of_type")
    print(f"We are adding IDs of type: {user_id_type}")

    staff_user_ids = get_config(f"custom.2i2c.staff_{user_id_type}_ids", [])
    print(f"Here are the IDs: {staff_user_ids}")

    c.Authenticator.admin_users.extend(staff_user_ids)
    print(f"Here are the admins: {c.Authenticator.admin_users}")
```

Logs:

```
Loading /usr/local/etc/jupyterhub/secret/values.yaml
No config at /usr/local/etc/jupyterhub/existing-secret/values.yaml
Loading extra config: 01-custom-theme
Loading extra config: 02-custom-admin
Loading extra config: 03-cloud-storage-bucket
Loading extra config: 04-2i2c-add-staff-user-ids-to-admin-users
We are adding staff IDs to the hub admin list!
Are we adding?: True
We made it to the if statement!
We are adding IDs of type: google
Here are the IDs: ['choldgraf@2i2c.org', 'colliand@2i2c.org', 'erik@2i2c.org', 'damianavila@2i2c.org', 'georgianaelena@2i2c.org', 'jmunroe@2i2c.org', 'sgibson@2i2c.org', 'yuvipanda@2i2c.org']
Here are the admins: <LazyConfigValue {'extend': ['choldgraf@2i2c.org', 'colliand@2i2c.org', 'erik@2i2c.org', 'damianavila@2i2c.org', 'georgianaelena@2i2c.org', 'jmunroe@2i2c.org', 'sgibson@2i2c.org', 'yuvipanda@2i2c.org']}>
```

GeorgianaElena commented:
@sgibson91, wondering if this is what I saw when testing out the temple hub [here](https://github.com/2i2c-org/infrastructure/pull/1664#issuecomment-1232752802)
which made me open https://github.com/2i2c-org/infrastructure/issues/1671.

Things kind of self-resolved back then and I couldn't identify the issue. But maybe the additional context there might help

sgibson91 commented:
Thanks @GeorgianaElena - yeah very strange. This doesn't seem to have self-rectified yet as of this morning.

sgibson91 commented:
I think this is a problem with Google IDs. I just deployed a dask-staging hub with CILogon with Google as the provider (the staging hub I deployed is using Auth0, also with Google) and I am not admin there either.

sgibson91 commented:
Hub logs from the latest deployment

```
Loading extra config: 04-2i2c-add-staff-user-ids-to-admin-users
We are adding staff IDs to the hub admin list!
Are we adding?: True
We made it to the if statement!
We are adding IDs of type: google
Here are the IDs: ['choldgraf@2i2c.org', 'colliand@2i2c.org', 'erik@2i2c.org', 'damianavila@2i2c.org', 'georgianaelena@2i2c.org', 'jmunroe@2i2c.org', 'sgibson@2i2c.org', 'yuvipanda@2i2c.org']
Here are the admins: <LazyConfigValue {'extend': ['choldgraf@2i2c.org', 'colliand@2i2c.org', 'erik@2i2c.org', 'damianavila@2i2c.org', 'georgianaelena@2i2c.org', 'jmunroe@2i2c.org', 'sgibson@2i2c.org', 'yuvipanda@2i2c.org']}>
```

researchdelight requires GitHub auth, so we'll see if the issue persists there (my prediction is that it won't)

sgibson91 commented:
I setup auth for the researchdelight hub the same way I did for the staging hub, via Auth0, but this time used GitHub IDs (as requested). **I am an admin on the research delight hub.**

So there's something about the Google IDs that cannot be made admin, or is failing when being made admin.

sgibson91 commented:
There is no specific mention of admin users in this issue, but I wonder if it's related?

* https://github.com/jupyterhub/oauthenticator/issues/399

sgibson91 commented:
Despite issues with admin access on the two staging hubs, I think this is now ready for review

sgibson91 commented:
@jmunroe Done!

jmunroe commented:
Is anything additional required on this cluster required to be able to support GPU enabled machines? (https://infrastructure.2i2c.org/en/latest/howto/features/gpu.html)

- [ ] Requesting Quota Increase
- [ ] Setup GPU nodegroup on eksctl
- [ ] Setting up a GPU user profile

For the user profile, using the exact same set up / images as is used in Pangeo I think makes sense:

```
jupyterhub:
   extraEnv:
      # Temporarily set for *all* pods, including pods without any GPUs,
      # to work around https://github.com/2i2c-org/infrastructure/issues/1530
      NVIDIA_DRIVER_CAPABILITIES: compute,utility
   singleuser:
      profileList:
        - display_name: Large + GPU
          description: 14GB RAM, 4 CPUs, T4 GPU
          profile_options:
            gpu:
            image:
              display_name: Image
              choices:
                tensorflow:
                  display_name: Pangeo Tensorflow ML Notebook
                  slug: "tensorflow"
                  kubespawner_override:
                    node.kubernetes.io/instance-type: g4dn.xlarge
                    image: "pangeo/ml-notebook:<tag>"
                pytorch:
                  display_name: Pangeo PyTorch ML Notebook
                  default: true
                  slug: "pytorch"
                  kubespawner_override:
                    node.kubernetes.io/instance-type: g4dn.xlarge
                    image: "pangeo/pytorch-notebook:<tag>"
          kubespawner_override:
            mem_limit: null
            mem_guarantee: 14G
            extra_resource_limits:
              nvidia.com/gpu: "1"
```
If a different instance-type is recommended, that this is just fine with me. The key things I want to be able to show:
a) we are able to spinup GPU enabled nodes
b) we are able use start different image for different user profiles

(If this is not an 'easy' ask, please let me know. Thanks.)

sgibson91 commented:
@jmunroe This should be doable - I will need to do some reading up on AWS-based GPUs, but we have docs and other examples for me to work from.

sgibson91 commented:
I've put in a quota increase request for 48 vCPUs, which should be 24 GPUs (quota goes on total number of vCPUs combined). Hopefully, if you're doing small, group demos this should be enough. But we can put another increase request in later if necessary.

sgibson91 commented:
Question for @yuvipanda: do we also create GPU node groups for the dask nodes, or only the notebook nodes?

sgibson91 commented:
I think docs build failure is unrelated?

yuvipanda commented:
@sgibson91 only the notebook nodes!

sgibson91 commented:
> * only deploy the reashdelight hub and remove the staging and dask-staging deployments (i believe they aren't actually needed, right?)

I actually don't know how our CI/CD will behave if we have a cluster that does not have a staging hub on it. So it might be needed for the sake of automatic deployment.

> * switch them all to GitHub based authentication (through cilogon, since I believe we still want to move away from auth0 eventually) and make sure the hubs are not open to everyone

I don't think we have a complete GitHub example in the docs here: https://infrastructure.2i2c.org/en/latest/hub-deployment-guide/configure-auth/cilogon.html#jupyterhub-cilogonoauthenticator I'm also not sure if we have a hub authenticating with GitHub via CILogon setup in the repo. A quick search didn't bring up anything. So I'd feel more comfortable using GitHub auth with a GitHub OAuth app directly, rather than via CILogon.

sgibson91 commented:
I figured out the issue!

Ok, I swapped both the staging and dask-staging hubs to GitHub auth, but found I was still not an admin. I deleted both deployments using `helm --namespace HUB_NAME delete HUB_NAME && kubectl delete namespace HUB_NAME` and then redeployed both hubs - still no admin!!

Then I realised that on `researchdelight` I had specifically added:

```yaml
admin_users:
  - jmunroe
```

I guess because I wanted to be _absolutely sure_ that he had admin powers. But I had not defined `admin_users` _at all_ for the staging or dask-staging hubs. I added the below config and redeployed:

```
admin_users: []
```

Boom! Admin rights!!

So I guess [this script](https://github.com/2i2c-org/infrastructure/blob/7073d647ba3cb38fdc174832d9c25a8e2a4af21c/helm-charts/basehub/values.yaml#L488-L495) is trying to append to a list that doesn't exist and is failing silently?

Anyway, incoming commits that change the staging and dask-staging configs

sgibson91 commented:
Omg, I did `git merge` ( not `git rebase`) so update my branch and I don't know why that has added so many commits instead of one 🤦🏻‍♀️

sgibson91 commented:
Ok, the dask-staging hub is now timing out "waiting for the condition" from helm and I don't know why because all the pods are up and running, so I'm just going to purge that one in view of getting this damn PR merged. staging hub completes fine.

sgibson91 commented:
Ugh, now the PR has changed 73 files, what the hell has happened 😭

sgibson91 commented:
Closing in favour of https://github.com/2i2c-org/infrastructure/pull/2022 which isn't so much of a mess
